src/main/java/fi/aalto/cs/apluscourses/utils/DomUtil.java

```diff
@@ -0,0 +1,121 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.xml.XMLConstants;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+public class DomUtil {
+
+  private static Logger logger = LoggerFactory.getLogger(DomUtil.class);
+
+  private static ConcurrentMap<String, XPathExpression> xPathCache;
+  private static final XPathFactory xPathFactory;
+  private static final DocumentBuilderFactory documentBuilderFactory;
+
+  static {
+    xPathCache = new ConcurrentHashMap<>();
+
+    xPathFactory = XPathFactory.newInstance();
+
+    documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    try {
+      documentBuilderFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+      documentBuilderFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
+    } catch (IllegalArgumentException e) {
+      logger.warn("Could not set XXE restrictions for DOM tools because the platform does not "
```

Comment:
[XXE defnition](https://www.acunetix.com/blog/articles/xml-external-entity-xxe-vulnerabilities/) :smiley_cat:

Reply:
Okay, to be honest, I am not an expert on these kind of security things. The XXE vulnerability (and how to fix it) was pointed out by Sonar. Do you think this solution is ok? (Both my Ubuntu machine and my Windows 10 machine at home fails to set those security measures thus falling to the catch block and logging the warning.)

As far as I understood it, this is not so crucial thing in our case, because users (students) should load XML files only from sources they trust (teachers), and as the result of the parse is not sent out from the users computer, there is no chance of information leaking. Only bad thing that could cause if a user parses a malicious XML (in this case *.iml) file they got from somewhere, would be their IDE to crash. Do you agree?

Reply:
I wasn't aware of this XXE thingy before, but I believe this definitely needs to stay!